internal/mutable/numericarray.go

```diff
@@ -0,0 +1,399 @@
+package mutable
+
+import (
+	"sync/atomic"
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/memory"
+)
+
+// Int64Array is an array of int64 values.
+type Int64Array struct {
+	refCount int64
+	mem      memory.Allocator
+	data     *memory.Buffer
+	rawData  []int64
+	length   int
+}
+
+// NewInt64Array constructs a new Int64Array.
+func NewInt64Array(mem memory.Allocator) *Int64Array {
+	return &Int64Array{
+		refCount: 1,
+		mem:      mem,
+	}
+}
+
+// Append will append a value to the array. This will increase
+// the length by 1 and may trigger a reallocation if the length
+// would go over the current capacity.
+func (b *Int64Array) Append(v int64) {
+	b.Reserve(1)
+	b.rawData = append(b.rawData, v)
+	b.length = len(b.rawData)
+}
+
+// AppendValues will append the given values to the array.
+// This will increase the length for the new values and may
+// trigger a reallocation if the length would go over the current
+// capacity.
+func (b *Int64Array) AppendValues(v []int64) {
+	b.Reserve(len(v))
+	b.rawData = append(b.rawData, v...)
+	b.length = len(b.rawData)
+}
+
+// Cap returns the capacity of the array.
+func (b *Int64Array) Cap() int { return cap(b.rawData) }
+
+// Len returns the length of the array.
+func (b *Int64Array) Len() int { return len(b.rawData) }
+
+// NewArray returns a new array from the data using NewInt64Array.
+func (b *Int64Array) NewArray() array.Interface {
+	return b.NewInt64Array()
+}
+
+// NewInt64Array will construct a new arrow array from the
+// buffered data.
+//
+// This will reset the current array.
+func (b *Int64Array) NewInt64Array() *array.Int64 {
+	data := array.NewData(
+		arrow.PrimitiveTypes.Int64,
+		len(b.rawData),
+		[]*memory.Buffer{nil, b.data},
+		nil, 0, 0,
+	)
+	b.reset()
+
+	a := array.NewInt64Data(data)
+	data.Release()
+	return a
+}
+
+func (b *Int64Array) init() {
+	b.data = memory.NewResizableBuffer(b.mem)
+}
+
+func (b *Int64Array) reset() {
+	b.data.Release()
+	b.data = nil
+	b.rawData = nil
+}
+
+// Retain will retain a reference to the builder.
+func (b *Int64Array) Retain() {
+	atomic.AddInt64(&b.refCount, 1)
+}
+
+// Release will release any reference to data buffers.
+func (b *Int64Array) Release() {
+	if atomic.AddInt64(&b.refCount, -1) == 0 {
+		if b.data != nil {
+			b.data.Release()
+			b.data = nil
+			b.rawData = nil
```

Comment:
why not `b.reset()`?

Reply:
Fixed. It's mostly because it didn't exist when I wrote this code. I've updated it to use reset and also wrote a test that `NewArray` causes the array to be reset.